internal/api/handler.go

```diff
@@ -3,47 +3,66 @@ package api
 import (
 	"context"
 	"fmt"
+	"html/template"
 	"net/http"
 	"net/url"
 
+	"github.com/gorilla/csrf"
 	"github.com/mattermost/chimera/internal/cache"
 
-	"github.com/gorilla/mux"
-	"github.com/mattermost/chimera/internal/providers"
+	"github.com/mattermost/chimera/internal/statuserr"
+	"github.com/pkg/errors"
+
 	"github.com/mattermost/chimera/internal/util"
 	"github.com/sirupsen/logrus"
 	"golang.org/x/oauth2"
 )
 
 const (
-	stateExtensionLength = 16
+	stateExtensionLength                   = 16
+	chimeraAuthorizationVerificationCookie = "chimera_authz_verification"
 )
 
-func NewHandler(appsRegistry map[string]OAuthApp, cache StateCache) *Handler {
-	return &Handler{
-		stateCache:   cache,
-		appsRegistry: appsRegistry,
+func NewHandler(cache StateCache, baseURL *url.URL, confirmFormPath, cancelPagePath string) (*Handler, error) {
+	confirmForm, err := template.ParseFiles(confirmFormPath)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to parse confirmation form template")
 	}
+
+	return &Handler{
+		stateCache:               cache,
+		baseURL:                  baseURL.String(),
+		confirmationFromTemplate: confirmForm,
+		cancelPagePath:           cancelPagePath,
+		useSecureCookie:          useSecureCookies(baseURL),
+	}, nil
 }
 
 type Handler struct {
-	stateCache   StateCache
-	appsRegistry map[string]OAuthApp
+	stateCache               StateCache
+	baseURL                  string
+	confirmationFromTemplate *template.Template
+	cancelPagePath           string
+	useSecureCookie          bool
 }
 
 type StateCache interface {
-	GetRedirectURI(state string) (string, error)
-	SetRedirectURI(state, redirectURI string) error
+	GetRedirectURI(state string) (cache.AuthorizationState, error)
+	SetRedirectURI(state string, authzState cache.AuthorizationState) error
+	DeleteState(state string) error
 }
 
-func (h *Handler) handleAuthorize(c *Context, w http.ResponseWriter, r *http.Request) {
-	app, found := h.getOAuthApp(c, r)
-	if !found {
-		w.WriteHeader(http.StatusNotFound)
-		return
-	}
-	c.Logger = c.Logger.WithField("provider", app.Provider).
-		WithField("app", app.Identifier)
+type AuthFormData struct {
+	RedirectURL    string
+	ConfirmAuthURL string
+	ProviderName   string
+	ProviderURL    string
+	CancelAuthURL  string
+	CsrfField      template.HTML
+}
+
+func (h *Handler) handleAuthorize(c *OAuthAppContext, w http.ResponseWriter, r *http.Request) {
+	c.Logger = loggerWithAppFields(c.Logger, c.OAuthApplication)
 	c.Logger.Infof("Handling authorization")
 
 	redirectURI := r.URL.Query().Get("redirect_uri")
```

Comment:
The URI scheme should be checked here and verified it's either `http` or `https`